pkg/server/plugin/nodeattestor/azureimds/imds.go

```diff
@@ -0,0 +1,416 @@
+		td:             coreConfig.TrustDomain,
+		tenants:        tenants,
+		idPathTemplate: tmpl,
+	}
+}
+
+type IMDSAttestorPlugin struct {
+	nodeattestorbase.Base
+	nodeattestorv1.UnsafeNodeAttestorServer
+	configv1.UnsafeConfigServer
+
+	log hclog.Logger
+
+	mu     sync.RWMutex
+	config *imdsAttestorConfig
+
+	hooks struct {
+		tenantIdMap         map[string]string
+		newClient           func(azcore.TokenCredential) (apiClient, error)
+		fetchCredential     func(string) (azcore.TokenCredential, error)
+		validateAttestedDoc func(context.Context, *azure.AttestedDocument) (*azure.AttestedDocumentContent, error)
+		lookupTenantID      func(string) (string, error)
+	}
+}
+
+var _ nodeattestorv1.NodeAttestorServer = (*IMDSAttestorPlugin)(nil)
+
+func New() *IMDSAttestorPlugin {
+	p := &IMDSAttestorPlugin{}
+	p.hooks.tenantIdMap = make(map[string]string)
+	p.hooks.newClient = newAzureClient
+	p.hooks.fetchCredential = func(tenantID string) (azcore.TokenCredential, error) {
+		return azidentity.NewDefaultAzureCredential(
+			&azidentity.DefaultAzureCredentialOptions{
+				TenantID: tenantID,
+			},
+		)
+	}
+	p.hooks.validateAttestedDoc = validateAttestedDocument
+	p.hooks.lookupTenantID = lookupTenantID
+
+	return p
+}
+
+func (p *IMDSAttestorPlugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+func (p *IMDSAttestorPlugin) Attest(stream nodeattestorv1.NodeAttestor_AttestServer) error {
+	// receive initial empty payload
+	if _, err := stream.Recv(); err != nil {
+		p.log.Info("received initial empty payload", "error", err)
+		return err
+	}
+
+	config, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	// create a 32byte nonce
+	nonce, err := generateRandomAlphanumeric(32)
+	if err != nil {
+		return err
+	}
+
+	// send nonce back to agent
+	if err := stream.Send(&nodeattestorv1.AttestResponse{
+		Response: &nodeattestorv1.AttestResponse_Challenge{
+			Challenge: []byte(nonce),
+		},
+	}); err != nil {
+		return err
+	}
+
+	// receive the attestation payload
+	req, err := stream.Recv()
+	if err != nil {
+		return err
+	}
+	// Get the challenge response which contains the attested document and metadata
+	payload := req.GetChallengeResponse()
+	if payload == nil {
+		return status.Error(codes.InvalidArgument, "missing attestation payload")
+	}
+
+	attestationData := new(azure.IMDSAttestationPayload)
+	if err := json.Unmarshal(payload, attestationData); err != nil {
+		return status.Errorf(codes.InvalidArgument, "failed to unmarshal data payload: %v", err)
+	}
+
+	// parse the document
+	docData, err := p.hooks.validateAttestedDoc(stream.Context(), &attestationData.Document)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "failed to validate attested document: %v", err)
+	}
+
+	switch {
+	case docData.VMID == "":
+		return status.Errorf(codes.InvalidArgument, "missing VM ID in attested document")
+	case validateUUID(docData.VMID) != nil:
+		return status.Errorf(codes.InvalidArgument, "invalid VM ID: %v", err)
```

Comment:
The err variable here is not related with this error. I think that this should be something like this:

```suggestion
	case err := validateUUID(docData.VMID); err != nil:
		return status.Errorf(codes.InvalidArgument, "invalid VM ID: %v", err)

```